pkg/tsdb/postgres/postgres.go

```diff
@@ -61,15 +86,161 @@ func escape(input string) string {
 	return strings.ReplaceAll(strings.ReplaceAll(input, `\`, `\\`), "'", `\'`)
 }
 
-func generateConnectionString(datasource *models.DataSource, logger log.Logger) (string, error) {
+type certFileType int
+
+const (
+	rootCert = iota
+	clientCert
+	clientKey
+)
+
+func (t certFileType) String() string {
+	switch t {
+	case rootCert:
+		return "root certificate"
+	case clientCert:
+		return "client certificate"
+	case clientKey:
+		return "client key"
+	default:
+		panic(fmt.Sprintf("Unrecognized certFileType %d", t))
+	}
+}
+
+// writeCertFile writes a certificate file.
+func (s *postgresService) writeCertFile(ds *models.DataSource, fileContent, dataDir string, fileType certFileType) error {
+	var jsonFieldName string
+	var filename string
+	switch fileType {
+	case rootCert:
+		jsonFieldName = "generatedTLSRootCertFile"
+		filename = "root.crt"
+	case clientCert:
+		jsonFieldName = "generatedTLSCertFile"
+		filename = "client.crt"
+	case clientKey:
+		jsonFieldName = "generatedTLSKeyFile"
+		filename = "client.key"
+	default:
+		panic(fmt.Sprintf("unrecognized certFileType %s", fileType.String()))
+	}
+
+	s.logger.Debug("Writing cert file", "type", fileType, "jsonFieldName", jsonFieldName)
+
+	var generatedFilePath string
+	if tlsjs, ok := ds.JsonData.CheckGet(jsonFieldName); ok {
+		// Get generated file path from data source
+		generatedFilePath = tlsjs.MustString("")
+		s.logger.Debug("Got file path from data source", "path", generatedFilePath)
+	} else {
+		s.logger.Debug("File path not stored in data source")
+	}
+	if fileContent != "" {
+		if generatedFilePath == "" {
+			// Use standard filename
+			generatedFilePath = filepath.Join(dataDir, filename)
+		}
+
+		s.logger.Debug("Writing cert file", "path", generatedFilePath)
+		if err := ioutil.WriteFile(generatedFilePath, []byte(fileContent), 0600); err != nil {
```

Comment:
I think that it would be good to remove any leading spaces (since the certificate will fail otherwise) and add a test for this case similar to this [one](https://github.com/grafana/grafana/blob/966078eb60979c87d18473f84a4a9f8257c41997/pkg/tsdb/postgres/postgres_test.go#L109).

Reply:
I tested with leading spaces, it seems pq package is removing spaces and it worked directly.

Reply:
If I understand correctly `pq` trims the spaces from the connection URL; what I meant was to trim the spaces from the input (certificate or key) before saving it in the file that will be provided to the connection string.
However I have tested again and my problem weer not the leading spaces but leading spaces/tabs on each line of the input which is an invalid input actually. So not bother with this, my bad!